### ISSUE TITLE
UniformWithReplacementSampler can return an empty batch

### ISSUE THREAD

**Issue by tholop** [pytorch/opacus#161]
Since UniformWithReplacementSampler is using Poisson subsampling, with some probability it can return an empty batch. This does not happen often, but when it happens it crashes the training loop because PyTorch's DataLoader expects a non-empty batch. 

https://github.com/pytorch/opacus/blob/5c01cf3e1bb42d647ebd418ebc610ceb966a0e83/opacus/utils/uniform_sampler.py#L42

![2021-03-31_Wed_09-27](https://user-images.githubusercontent.com/32133147/113151971-a1ada100-9203-11eb-9e90-85505021260c.png)

If my understanding is correct, we could fix this by either:
1. Modifying UniformWithReplacementSampler to check whether `indices` is empty before we return it, and resample (e.g. repeat the Poisson subsampling, or return a batch of size 1 by choosing one element uniformly at random). It is slightly inaccurate from a privacy accounting perspective though (but the difference is tiny for usual batch sizes)
2. Modifying the training/testing loop of all the examples to take into account the fact that some batches can be empty (maybe with a different data loader)

I'm happy to send a PR

**Comment by alexandresablayrolles** [comment]
Thanks for raising this issue! As you said, 1) would result in inaccurate accounting and the goal of Poisson sampling is to really match the theoretical guarantees. 
We're working on implementing a variant of 2), we will update the issue when the PR is around (probably by next week).

**Comment by alexandresablayrolles** [comment]
PR https://github.com/pytorch/opacus/pull/164 if you want to try it out

**Comment by tholop** [comment]
Thanks, this sampling for the number of empty batches is neat. Just one question regarding the new `step()` function: why should we take `batch_size = 1` to calibrate the noise on empty batches? https://github.com/alexandresablayrolles/opacus/blob/f207de1cf95203b4e67acc451f1f5a547d8fdfb3/opacus/privacy_engine.py#L401

**Comment by mmalekzadeh** [comment]
@alexandresablayrolles, I enjoyed reading this PR and learned new things.
 
Just some suggestions:
1. Better to document  "is_empty"  parameter added to `step(self, is_empty: bool=False)` in the description of  mthod.
2. As user can input a list for `max_grad_norm`  (and not just a float), I was wondering maybe this line `clip_values = [self.max_grad_norm for p in self.module.parameters() if p.requires_grad]` could provide an error. Cause it generates a list of lists. Not sure though.

Overall, it is worth elaborating a bit more on this solution. It seems to me that this solution estimates the number of empty batches (e.g. K) and at the beginning it perform K fake updates. Like randomly changing the randomly initialized model K times and then start learning. Is it right?!

Thanks,

**Comment by alexandresablayrolles** [comment]
@tholop: good catch, I guess we have to use the expected batch size in this. I will also change the gradient computation for loss_reduction='mean' to use the expected batch size instead of the current batch size. 

@mmalekzadeh: good points 1) and 2), will update it in the PR. In this solution, we estimate the number of empty batches at the beginning and after every non-empty batch.
